devito/arguments.py

```diff
@@ -158,198 +145,398 @@ class PtrArgument(Argument):
+    def visit_DimensionParameter(self, o):
+        dependency = ValueDependency(o)
+        size = ScalarArgument(o.provider.size_name, [dependency])
+        start = ScalarArgument(o.provider.start_name, [dependency])
+        end = ScalarArgument(o.provider.end_name, [dependency])
+        return [size, start, end]
 
-    def reset(self):
-        for i in self.rtargs:
-            i.reset()
+    def visit_Object(self, o):
+        return PtrArgument(o)
 
-    @property
-    def value(self):
-        """ Returns a tuple (same order as rtargs) with the current value of each rtarg
-            If any of the rtargs has value None, the return value here is None.
-        """
-        child_values = tuple([i.value for i in self.rtargs])
-        return child_values if all(i is not None for i in child_values) else None
+    def visit_Array(self, o):
+        return TensorArgument(o)
 
-    @property
-    def dtype(self):
-        """The data type of the iteration variable"""
-        return np.int32
-
-    @cached_property
-    def rtargs(self):
-        size = ScalarArgument(self.size_name, self, max)
-        start = ScalarArgument(self.start_name, self, max, 0)
-        end = ScalarArgument(self.end_name, self, max)
-        return namedtuple("RuntimeArguments", ["size", "start", "end"])(size, start, end)
-
-    # TODO: Can we do without a verify on a dimension?
-    def verify(self, value, enforce=False):
-        verify = True
-        if value is None:
-            if self.value is not None:
-                return True
-
-            try:
-                parent_value = self.parent.value
-                if parent_value is None:
-                    return False
-            except AttributeError:
-                return False
-        # Make sure we're dealing with a 3-tuple.
-        # See docstring of infer_dimension_values_tuple for more info
-        value = infer_dimension_values_tuple(value, self.rtargs)
-        if hasattr(self, 'parent'):
-            parent_value = self.parent.value
-            if parent_value is not None and not enforce:
-                parent_value = infer_dimension_values_tuple(parent_value, self.rtargs)
-                value = tuple([self.reducer(i1, i2) for i1, i2 in zip(value,
-                                                                      parent_value)])
-            verify = verify and self.parent.verify(value)
-        if value == self.value:
-            return True
-
-        # Derived dimensions could be linked through constraints
-        # At this point, a constraint needs to be added that enforces
-        # dim_e - dim_s < SOME_MAX
-        # Also need a default constraint that dim_e > dim_s (or vice-versa)
-        verify = verify and all([a.verify(v, enforce=enforce) for a, v in
-                                 zip(self.rtargs, value)])
-        return verify
+    def visit_Scalar(self, o):
+        arg = ScalarArgument(o.name, o, dtype=o.dtype)
+        arg.provider = o
```

Comment:
Nope. You're creating a ``ScalarArgument`` and then, externally, you're dynamically adding an attribute to it. Use the constructor instead.

Reply:
Yes, because according to my design, `ScalarArgument`s don't need a provider but in `yask/operator.py/arguments` your code assumes that they have `provider`s. I'm only adding the provider for compatibility with your code since I want to limit the amount of change this PR brings. In a future PR I will clean up your code so it doesn't need a `provider` on `ScalarArgument`s.

Reply:
No. This is seriously a bad hack. I accept postponing all other refactorings, but this should really really go.